Automattic/wp-calypso
# Add contextual trigger for Backup add-on products ( with dynamic add-on suggestion )

This is the follow-up of PR #69377 with add-on suggested dynamically 

#### Proposed Changes

* Currently add on suggestion in the previous PR goes like this: 10GB -> 10GB and 1TB -> 1TB 
* This PR uses the same logic only when overall usage is less than the total storage
* If usage is above the total storage of the base plan, we dynamically calculates the suggestion as follows
  * The additional storage needed is calculated as **lapsed storage + 25% of lapsed storage**  i.e if base plan is 10GB and the user has used around 30GB then additional storage needed is **( (30GB - 10GB)  + 25%(30GB - 10GB)) = 25GB**
  * After determining the needed storage (as above) see if it is above 1TB (since 1TB is our maximum upgrade available). Simply suggest 1TB Add-on to the user
  * Else from the monotonically increasing array i.e [10GB, 100GB, 1TB] find the first one above the needed storage and suggest the same to the user
  
#### Testing Instructions

* Create a JN site with Jetpack Backup 10GB product (or) Jetpack Security 10GB
*  click on Jetpack Cloud live link below
    * Goto `/backup`
* or boot up this PR 
    * Run `git fetch && git checkout add/dynamic-backup-addon`
    * Run `yarn start-jetpack-cloud`
    * Goto http://jetpack.cloud.localhost:3000/backup
* Validate the following scenarios
  * WIth 10GB base plan and usage less than 10GB (For ex: 6.5GB), the addon suggested should be 10GB
    <img width="742" alt="Screenshot 2022-10-24 at 6 29 29 PM" src="https://user-images.githubusercontent.com/2027003/197536870-a2b37b79-e2a5-40ad-a850-c478a88751d5.png">
  
  * With 1TB Base plan (launch a new JN site with 1TB) and usage is less than 1TB (For ex: 838 GB), the addon suggested should be 1TB
    <img width="745" alt="Screenshot 2022-10-26 at 1 32 38 AM 2" src="https://user-images.githubusercontent.com/2027003/197870741-c2b8327c-6ad8-41d1-88a8-c7d56d04dcfc.png">

  * With 1TB Base plan and usage is more than 1TB (For Ex: 8381GB), the addon suggested should be 1TB
     <img width="731" alt="Screenshot 2022-10-26 at 1 35 05 AM" src="https://user-images.githubusercontent.com/2027003/197871206-ce5ff219-076f-458f-9107-faf1e8435d92.png">

  * With 10GB Base plan and usage is more than 10GB but less than 20GB (for Ex: 14GB)
     <img width="730" alt="Screenshot 2022-10-26 at 1 44 37 AM" src="https://user-images.githubusercontent.com/2027003/197872803-ca45a162-f592-48a5-a450-a1b7292ee925.png">

  * With 10GB Base plan and usage is more than 20GB (for Ex: 24GB)
     <img width="728" alt="Screenshot 2022-10-26 at 1 46 53 AM" src="https://user-images.githubusercontent.com/2027003/197873203-5cdfffa2-3214-49e0-8828-1de98c429d4e.png">

  * With 10GB Base plan and usage is more than 100GB (for Ex: 149GB)
    <img width="729" alt="Screenshot 2022-10-26 at 6 30 35 PM" src="https://user-images.githubusercontent.com/2027003/198032460-b721896b-4931-46d6-bcc4-b877265d4176.png">



**TIP:** You can simulate different storage usages with Redux DevTools in your browser. Dispatch the following action, substituting `<your_site_id>` with your site's numeric ID and `<bytes>` with a number of bytes used:

(For reference: 6GB = `6900000000`; 838GB = `899000000000`; 8381GB = `8990000000000` 14GB = `14999999999` 24GB= `25999999999` 149GB=`159999999999`)

```js
{
  type: 'REWIND_SIZE_SET',
  siteId: <your_site_id>,
  size: {
    bytesUsed: <bytes>
  }
}
```

#### Pre-merge Checklist






Related to #
1203228880371291-as-1203228880371316/f
